astropy/cosmology/_src/signature_deprecations.c

```diff
@@ -27,13 +27,17 @@ depr_kws_wrap_dealloc(DeprKwsObject* self)
     Py_XDECREF(self->wrapped);
     Py_XDECREF(self->names);
     Py_XDECREF(self->since);
-    Py_TYPE(self)->tp_free((PyObject*)self);
+    PyTypeObject* type = Py_TYPE((PyObject*)self);
+    freefunc free_func = PyType_GetSlot(type, Py_tp_free);
+    free_func((PyObject*)self);
+    Py_DECREF((PyObject*)type);
```

Comment:
Like in the other PR, as I understand it `Py_TYPE` returns a borrowed reference, so this should not be `DECREF`d.

But then I see that @da-woods recommended it. So, this leaves me confused...

Reply:
If I followed correctly, you were hinting at a similar discussion in #18231, which was resolved as "no action needed".
My conclusion would be then: can we merge this one now ?

Reply:
Yes, I was wrong there and here, so resolving this conversation.